lib/AST/Type.cpp

```diff
@@ -5213,7 +5228,13 @@ AnyFunctionType *AnyFunctionType::getAutoDiffDerivativeFunctionLinearMapType(
     for (auto diffParam : diffParams) {
       auto paramType = diffParam.getPlainType();
       auto paramTan = paramType->getAutoDiffTangentSpace(lookupConformance);
-      assert(paramTan && "Parameter has no tangent space?");
+      // Error if paraneter has no tangent space.
+      if (!paramTan) {
+        return llvm::make_error<DerivativeFunctionTypeError>(
+            this,
+            DerivativeFunctionTypeError::Kind::NonDifferentiableParameters,
+            parameterIndices);
```

Comment:
The `IndexSubset *` passed here is not correct: it should be the non-differentiable parameter indices, not all parameter indices.

I guess this error doesn't trigger because attribute type-checking performs some upstream validation.

In a follow-up, I plan to rewrite this error payload to just be the parameter `Type`, just like `DerivativeFunctionTypeError::Kind::NonDifferentiableResult`.

Reply:
Fixed in https://github.com/apple/swift/pull/31755. The payload is now `std::pair<Type, unsigned>`.